pyads/pyads_ex.py

```diff
@@ -149,6 +149,85 @@ def adsAddRoute(net_id, ip_address):
+    data_header += (0x0500).to_bytes(2, 'big')						# Write command
+    data_header += (0x00000c00).to_bytes(4, 'big')					# Block of unknown
+    data_header += len(sending_host_name).to_bytes(2, 'little')		# Length of sender host name
+    data_header += sending_host_name.encode('utf-8')				# Sender host name
+    data_header += (0x0700).to_bytes(2, 'big')						# Block of unknown
+
+
+    actual_data = (6).to_bytes(2, 'little')							# Byte length of AMS ID (always 6)
+    actual_data += bytes(map(int, added_net_id.split('.')))			# Net ID being added to the PLC
+    actual_data += (0x0d00).to_bytes(2, 'big')						# Block of unknown (maybe encryption?)
+    actual_data += len(username).to_bytes(2, 'little')				# Length of the user name field
+    actual_data +=	username.encode('utf-8')						# PLC Username
+    actual_data += (0x0200).to_bytes(2, 'big')						# Block of unknown
+    actual_data += len(password).to_bytes(2, 'little')				# Length of password field
+    actual_data += password.encode('utf-8')							# PLC Password
+    actual_data += (0x0500).to_bytes(2, 'big')						# Block of unknown
+    actual_data += len(route_name).to_bytes(2, 'little')			# Length of route name
+    actual_data += route_name.encode('utf-8')						# Name of route being added to the PLC
+
+    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock: # UDP
+        # Listen on 55189 for the response from the PLC
+        sock.bind(('', 55189))
+
+        # Send our data to 48899 on the PLC
+        sock.sendto(data_header + actual_data, (ip_address, 48899))
+
+        # Response should come in in less than .5 seconds, but wait longer to account for slow communications
+        sock.settimeout(5)
+        # Allow TimeoutError to be raised so user can handle it how they please
+
+        # Keep looping until we get a response from our PLC
+        addr = [0]
+        while addr[0] != ip_address:
+            data, addr = sock.recvfrom(32) # PLC response is 32 bytes long
+
+    rcvd_packet_header = data[0:12]							# AMS Packet header, seems to define communication type
+    # If the last byte in the header is 0x80, then this is a response to our request
+    if rcvd_packet_header[-1] == 0x80:
+        rcvd_PLC_AMS_ID = data[12:18]						# AMS ID of PLC
+        # Convert to a String AMS ID
+        #rcvd_PLC_AMS_ID = '.'.join([str(int.from_bytes(rcvd_PLC_AMS_ID[i:i+1], 'big')) for i in range(0, len(rcvd_PLC_AMS_ID), 1)])
+        rcvd_AMS_port = data[18:20]							# Some sort of AMS port? Little endian
+        rcvd_command_code = data[20:22]						# Command code (should be read) Little endian
+        rcvd_protocol_block = data[22:]						# Unknown block of protocol
+        rcvd_is_password_correct = rcvd_protocol_block[4:7]	# 0x040000 when password was correct, 0x000407 when it was incorrect
+   
+    if int.from_bytes(rcvd_is_password_correct, 'big') == 0x040000:
+        return True
+    else:
```

Comment:
If value is 0x040000 when password was correct and 0x000407 when it was incorrect, we must test both
and return an exception if none of this values are read

Reply:
a34d4d596f185f0e5a26aa8b842b4c5c74403333